File: scenarios/csharp/dotnetcore/call-center/call-center/RestHelper.cs
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

using System;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;

namespace CallCenter
{
    public static class RestHelper
    {
        public static async Task<(HttpResponseMessage response, string content)> SendGet(string uri, string key, HttpStatusCode[] expectedStatusCodes)
        {
            using (var client = new HttpClient())
            using (var request = new HttpRequestMessage())
            {
                request.Method = HttpMethod.Get;
                request.RequestUri = new Uri(uri);
                request.Headers.Add("Ocp-Apim-Subscription-Key", key);

                var response_1 = await client.SendAsync(request);
                if (response_1.Content == null)
                {
                    throw new Exception($"The response from {uri} contains no content.");
                }
                var response_2 = await response_1.Content.ReadAsStringAsync();
                if (!expectedStatusCodes.Contains(response_1.StatusCode))
                {
                    throw new Exception($"The response from {uri} has an unexpected status code: {response_1.StatusCode}");
                }
                
                return (response_1, response_2);
            }
        }

        public static async Task<(HttpResponseMessage response, string content)> SendPost(string uri, string requestBody, string key, HttpStatusCode[] expectedStatusCodes)
        {
            using (var client = new HttpClient())
            using (var request = new HttpRequestMessage())
            {
                request.Method = HttpMethod.Post;
                request.RequestUri = new Uri(uri);
                request.Headers.Add("Ocp-Apim-Subscription-Key", key);

                if (!String.IsNullOrEmpty(requestBody))
                {
                    request.Content = new StringContent(requestBody, Encoding.UTF8, "application/json");
                }
                var response_1 = await client.SendAsync(request);
                if (response_1.Content == null)
                {
                    throw new Exception($"The response from {uri} contains no content.");
                }
                var response_2 = await response_1.Content.ReadAsStringAsync();
                if (!expectedStatusCodes.Contains(response_1.StatusCode))
                {
                    throw new Exception($"The response from {uri} has an unexpected status code: {response_1.StatusCode}. Response:{Environment.NewLine}{response_2}");
                }
                
                return (response_1, response_2);
            }
        }

        public static async Task SendDelete(string uri, string key)
        {
            using (var client = new HttpClient())
            using (var request = new HttpRequestMessage())
            {
                request.Method = HttpMethod.Delete;
                request.RequestUri = new Uri(uri);
                request.Headers.Add("Ocp-Apim-Subscription-Key", key);

                var response = await client.SendAsync(request);
                if (HttpStatusCode.NoContent != response.StatusCode)
                {
                    throw new Exception($"HTTP DELETE request returned unexpected status code: {response.StatusCode}. Full response:{Environment.NewLine}{response.ToString()}");
                }
                return;
            }
        }
    }
}

File: scenarios/python/console/call-center/rest_helper.py
#
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
#

# To install, run:
# python -m pip install requests
import requests
from typing import Dict, List

def send_get(uri : str, key : str, expected_status_codes : List[int]) -> Dict :
    headers = {"Ocp-Apim-Subscription-Key": key}
    response = requests.get(uri, headers=headers)
    if response.status_code not in expected_status_codes :
        raise Exception(f"The GET request to {uri} returned a status code {response.status_code} that was not in the expected status codes: {expected_status_codes}")
    else :
        try :
            # response.json() throws if the response is empty.
            response_json = response.json()
            return { "headers" : response.headers, "text" : response.text, "json" : response_json }
        except Exception :
            return { "headers" : response.headers, "text" : response.text, "json" : None }

def send_post(uri : str, content : Dict, key : str, expected_status_codes : List[int]) -> Dict :
    headers = {"Ocp-Apim-Subscription-Key": key}
    response = requests.post(uri, headers=headers, json=content)
    if response.status_code not in expected_status_codes :
        raise Exception(f"The POST request to {uri} returned a status code {response.status_code} that was not in the expected status codes: {expected_status_codes}")
    else :
        try :
            response_json = response.json()
            return { "headers" : response.headers, "text" : response.text, "json" : response_json }
        except Exception :
            return { "headers" : response.headers, "text" : response.text, "json" : None }

def send_delete(uri : str, key : str, expected_status_codes : List[int]) -> None :
    headers = {"Ocp-Apim-Subscription-Key": key}
    response = requests.delete(uri, headers=headers)
    if response.status_code not in expected_status_codes :
        raise Exception(f"The DELETE request to {uri} returned a status code {response.status_code} that was not in the expected status codes: {expected_status_codes}")
